backup_swift/tasks.py

```diff
@@ -0,0 +1,120 @@
+#
+"""
+Tarsnap utilities
+"""
+
+# Imports #####################################################################
+
+import datetime
+import logging
+import requests
+
+from django.conf import settings
+from django.core.mail import mail_admins
+from huey.api import crontab
+from huey.contrib.djhuey import db_periodic_task, db_task
+from swiftclient.service import SwiftError
+
+from backup_swift.tarsnap import make_tarsnap_backup
+
+from instance import openstack
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+# Functions #####################################################################
+
+
+def ping_heartbeat_url(url):
+    """
+    Pings heartbeat url.
+    :param str url: Url to ping
+    :return: True if http request finished and returned proper status, False otherwise.
+    """
+    try:
+        response = requests.get(url, timeout=30)
+        return 200 <= response.status_code < 300
+    except requests.RequestException:
+        return False
+
+
+def do_backup_swift():
+    """
+    Perform full swift backup sequence.
+    """
+
+    logger.info("Starting backup of swift containers")
+    error_report = ""
+    download_results = []
+    try:
+        download_results = openstack.download_swift_account(settings.BACKUP_SWIFT_TARGET)
+    except SwiftError:
+        error_report += "Miscellaneous error while downloading swift containers\n"
+        logger.exception("Misc error while downloading swift containers")
+
+    if download_results:
+        error_report += "Following containers failed to download:\n"
+        for container in download_results:
+            error_report += "#. {name}; Failed files: {count}.\n".format(
+                name=container.name, count=container.number_of_failures
+            )
+
+    # In case of downloading errors run tarsnap nevertheless, so we at least backup something.
+
+    tarsnap_succesfull = make_tarsnap_backup(
+        keyfile=settings.BACKUP_SWIFT_TARSNAP_KEY_LOCATION,
+        cachedir=settings.BACKUP_SWIFT_TARSNAP_CACHE_LOCATION,
+        directory=settings.BACKUP_SWIFT_TARGET,
+        archive_name="{}-{}".format(settings.BACKUP_SWIFT_TARSNAP_KEY_ARCHIVE_NAME, datetime.datetime.now().isoformat())
+    )
+
+    if not tarsnap_succesfull:
+        error_report += "Error while running tarsnap\n"
+
+    if error_report:
+        mail_admins("Error when backing up swift containers", error_report)
+        logger.error("Error when backing up swift containers %s", error_report)
+    else:
+        logger.info("Swift backup finished successfully")
+        if settings.BACKUP_SWIFT_SNITCH:
+            ping_heartbeat_url(settings.BACKUP_SWIFT_SNITCH)
+
+
+@db_task()
+def backup_swift_task():
+    """
+    Task that performs backup of swift containers.
+    """
+    do_backup_swift()
+
+
+if settings.BACKUP_SWIFT_ENABLED:
+
+    @db_periodic_task(crontab(minute="10", hour="1"))
```

Comment:
Ech. Looks like every single task in IM uses database --- to log to database. I'll change the decorators back.